### PR TITLE
Fix a race condition on _start_other

### PR DESCRIPTION
**Pull request by nategraff-sifive**
This might not be the best fix for this race condition, but it does fix it.

Track the need for a better solution here: [freedom-metal#102](https://github.com/sifive/freedom-metal/issues/102).

### CODE REVIEW

File: multicore-hello.c [hunk 1]
```diff
@@ -30,14 +30,25 @@ int secondary_main(void) {
...
 		int rc = metal_lock_init(&my_lock);
 		if(rc != 0) {
-			printf("Failed to initialize my_lock: %d", rc);
+			puts("Failed to initialize my_lock\n");
 			exit(1);
 		}
 
+		/* Ensure that the lock is initialized before any readers of
+		 * _start_other */
+		__asm__ ("fence rw,w"); /* Release semantics */
```
Comment (ingallsj) [inline multicore-hello.c:39]: 👍

**Comment by hcook** [comment]
I agree this could probably be handled better by the startup code, but your assumption about the relative timing of the harts seems sound in every system we have built so far.

**Review by hcook — APPROVED** [review]

File: multicore-hello.c [hunk 2]
```diff
@@ -30,14 +30,25 @@ int secondary_main(void) {
...
 		_start_other = 1;
 
 		return main();
 	} else {
+		/* We can't count on hart 0 to initialize _start_other before
+		 * the other harts get here because hart 0 is in charge of copying
+		 * the data section, zeroing bss, and running the constructors.
+		 * The other harts will always zero _start_other here long before
+		 * hart 0 gets around to setting it to 1 */
+		_start_other = 0;
```
Comment (ingallsj) [inline multicore-hello.c:50]: I thought that this was done by initializing "volatile int _start_other = 0;"?
I worry that this is a non-deterministic race, and we are relying on harts >0 winning the race before hart 0.

Comment (nategraff-sifive) [reply multicore-hello.c:50]: We are relying on that, but hart 0 has a lot of work to do before it gets here that the other harts don't. I expect it's something we can get away with in the short-term so I created a tracking issue in Freedom Metal to note that we need a better solution.

**Pull request merged**